src/base/Gas.ml

```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+                        (p, expr_static_cost e')) ),
+                erep ) )
+      | Builtin _
+      (* We don't add costs for Builtin because we can't know it statically
+       * without type info (Eval doesn't run the type checker). To know this
+       * statically, call builtin_cost below, providing argument types. *)
+      | GasExpr _ ->
+          ee
+    in
+    (e', erep)
 
   (* this is a dynamic cost. *)
+  let rec stmts_cost = function
+    | [] -> []
+    | (s, srep) :: rem_stmts ->
+        let s' =
+          match s with
+          | Load (x, _) ->
+              let g =
+                GasStmt
+                  (GasCharge.SumOf
+                     ( GasCharge.SizeOf (GI.get_id x),
+                       GasCharge.MapSortCost (GI.get_id x) ))
+              in
+              (* We charge *after* the load because we can't know the size before. *)
+              [ (s, srep); (g, srep) ]
+          | Store (_, v) | SendMsgs v | CreateEvnt v ->
+              let g = GasStmt (GasCharge.SizeOf (GI.get_id v)) in
+              [ (g, srep); (s, srep) ]
+          | Bind (x, e) ->
+            let g = GasStmt (GasCharge.StaticCost 1) in
+            let s' = Bind (x, expr_static_cost e) in
+            [ (g, srep); (s', srep) ]
+          | ReadFromBC _ | CallProc _ ->
+              let g = GasStmt (GasCharge.StaticCost 1) in
+              [ (g, srep); (s, srep) ]
+          | MapUpdate (_, klist, ropt) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                match ropt with
+                | Some r -> GasCharge.SumOf (GasCharge.SizeOf (GI.get_id r), n)
+                | None -> n
+              in
+              [ (GasStmt g, srep); (s, srep) ]
+          | MapGet (x, _, klist, _) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                  GasCharge.SumOf
+                    ( GasCharge.SumOf
+                        ( GasCharge.SizeOf (GI.get_id x),
+                          GasCharge.MapSortCost (GI.get_id x) ),
+                      n )
+              in
+              [ (s, srep); (GasStmt g, srep) ]
+          | MatchStmt (x, clauses) ->
+              let g = GasCharge.StaticCost (List.length clauses) in
+              let clauses' =
+                List.map clauses ~f:(fun (p, stmts) ->
+                    let stmts' = stmts_cost stmts in
+                    (p, stmts'))
+              in
+              let s' = MatchStmt (x, clauses') in
+              [ (GasStmt g, srep); (s', srep) ]
+          | AcceptPayment ->
+              let g = GasStmt (GasCharge.StaticCost 1) in
+              [ (g, srep); (s, srep) ]
+          | Iterate (l, _) ->
+              let g = GasStmt (GasCharge.LengthOf (GI.get_id l)) in
+              [ (g, srep); (s, srep) ]
+          | Throw _ (* TODO: Throw should charge same as event and send. *)
+          | GasStmt _ ->
+              [ (s, srep) ]
```

Comment:
I think this should throw an error. The funtion adds gas charge statements before or after the statement for which gas is charged, so if there is already a gas charge statement in the AST, then something has gone wrong.

Reply:
There are two reasons I left it this way:

1. Handling errors here (monadically) has a domino effect up until Runner.ml, and that means I'll have to manually check for errors at multiple places there (making it a bit ugly in Runner.ml).
2. More importantly, I wanted to keep it this way so that in case we introduce explicit gas charge statements / expressions in the language (though I can't think of a reason for having such a thing right now), this will just work.

But having said that, I am flexible about this, and if you think it's better to do what you said, I'll do it.

Reply:
1. You don't need to handle errors here. You just need to raise an error. It would be an internal error because this branch shouldn't happen.
2. If we want to add explicit gas charges, then this branch should be changed, but until we do we would expect this to be an error. And at that point we will discover that this branch should be changed, because we will hit it. I don't see any reason to not throw an error here.